### ISSUE TITLE
Limit API calls client side

### ISSUE THREAD

**Issue by clburlison** [clburlison/pinpoint#19]
Per @erikng: Have the clients poll how many API requests have been submitted in a day and not run if it's over 1,000.

**Comment by erikng** [comment]
Sneaky devil! Or is it 2,500?

**Comment by clburlison** [comment]
https://github.com/clburlison/pinpoint/wiki#google-geocoding-api :P

**Comment by erikng** [comment]
2,500 it is then!

**Comment by clburlison** [comment]
On more thinking...this seems redundant to another feature that was included with #4.

See commit https://github.com/clburlison/pinpoint/commit/6d23be5765dca54ef496fdb4e25b42a9cc7e0070#diff-9cee2e349b1920be4296162259506b00R131 the `check_wait_period()` function. 

Which that change by default pinpoint will only run once ever ~30min. With that it should only run 48 times per day which is way under the limit. 

No issue writing more code, but does it solve anything?

**Comment by erikng** [comment]
Let's say 50 for easy math. Wouldn't that be limited to 50 machines per day?

**Comment by clburlison** [comment]
> Wouldn't that be limited to 50 machines per day?

Yes. 

Your actual request is checking with Google to see if you are about to go over the limit. I'm not actual sure if that is do-able. This isn't exposed when you are using the free service, it might be doable if we included an optional API key. This entire request not even be worth my time... due to issue #2

Apple seems to not have a hard limit and if so it isn't publicly posted anywhere. Seeing as how we are already limiting this with the `check_wait_period()` function I doubt even a large fleet would hit that (hypothetical) limit.

As far as I'm concerned this type of request is the same as a lookup from the Maps.app so there very well might not be a limit if I can do this with the Geocoder from CoreLocation.

**Comment by erikng** [comment]
I agree with your thoughts. If Apple's API is less restrictive, your time is much better suited working on that. 

I will send you some other thoughts in a PM.

**Comment by flammable** [comment]
Hey Clayton!

Thanks again for this!  My concern is similar to Erik's - if I have a fleet of 715 Macs (and growing), should I hold off from deploying this to every machine?  The impression that I get is that each Mac does its own lookups, so the 2,500 limit is per-machine - not 2,500 requests for my lone MunkiReport server.

Assuming the API limit is per-machine...each Mac on our network has a unique IP address, so unless Google decides to ban our entire IP block, I think we're okay.  Admins that use NAT might need to be more careful.

I'm really excited to use this, but want to make sure I don't upset Google. :)

Mike

**Comment by erikng** [comment]
https://developers.google.com/maps/articles/geocodestrat#client

Is it possible to send the geolocation data only and then render it during the MR-PHP page view? We would lose the ability to see all the pins everywhere, but that would significantly reduce the hits per day.

**Comment by clburlison** [comment]
Still in class so I'll update fully later however in case it isn't obvious - I'm doing a call out to Google for one thing...Reverse geocoding to resolve `37.332112,-122.0329646` to Infinite Loop, Cupertino, CA 95014.

This can be disabled with the `AddressLookup` key https://github.com/clburlison/pinpoint/wiki/Preferences#supported-keys.

Munkireport will still fully function with the AddressLookup disabled. MR does the drawing of the maps, and placement of the pins based of the geo-cords.

**Comment by erikng** [comment]
May have found a EULA issue. 

https://www.google.com/work/mapsearth/products/mapsapi.html

```Standard Plan
The Google Maps APIs Standard Plan allows you to pay for uplifts above free usage limits for individual Maps APIs. The Standard Plan can be used for free, external, and publicly available implementations. Your service must be freely and publicly accessible to end users.
```

**Comment by flammable** [comment]
Oh cool.  So, if I disable `AddressLookup`, none of the clients will contact Google at all?

That'd be a decent workaround for the API limit.

**Comment by erikng** [comment]
Interestingly enough, this API limit is 100,000 per day.

**Comment by clburlison** [comment]
@flammable 

1. Yes, I would recommend disabling the `AddressLookup` if you're concerned about the API limit. Honestly the docs on how Google's APIs are spread out and confusing. The "Maps" API alone has at least 10 different doc locations.
1. As for how the service is counting 2,500 requests is also up for debate. I can't find a clear answer on any of those docs. My assumption is that it is based off your external IP address. I'd prefer to not blacklist my home address as a test. :grinning: 
1. If your goal is to help show Machine X is in building C this should be perfect. From my tests, CoreLocation is able to obtain a location accurate enough to determine the building I am in. (pic below)


@erikng 

1. Your EULA reference above is for a different section of the Map API. That deals with usage of maps/earth inside of websites/apps. I believe this to be the reference you want: https://developers.google.com/maps/terms#10-license-restrictions

 > Is it possible to send the geolocation data only and then render it during the MR-PHP page view? We would lose the ability to see all the pins everywhere, but that would significantly reduce the hits per day.

1. Yes, that's already how MunkiReport works, disable the reverse lookup with the `AddressLookup` key.

As far as I can tell this project and the implementation within MR are fully within the guidelines of Google's API usage terms. Even with MR being password protected a section of the term allows for this. If we were using MR as a third party and selling as a service provider, like Sal+, then we would _have_ to use a paid API account. Even that might still not be under the terms.

---

Lastly, for reference when you do disable the `AddressLookup` key this is the end result in MR.

![mr_global](https://cloud.githubusercontent.com/assets/2831320/13625908/2d64a7a4-e582-11e5-8555-9217b9ba09ef.png)
![mr_single](https://cloud.githubusercontent.com/assets/2831320/13625907/2d643954-e582-11e5-9488-5d0bf9fa358b.png)


If you wanted to do a reverse lookup at this point you can simply copy/paste the coordinates from the MR listing into Google Maps to obtain an estimate.

**Comment by flammable** [comment]
@clburlison That is perfect. Thank you so much!

**Comment by clburlison** [comment]
This was an impossible goal (due to the technology). However, it does bring up some good conversations about using Google's geocoding service. Since Apple's GeoCoder has been implemented in #2 I'm closing this.
